### PR TITLE
[sungrow] Initial contribution

### PR DESCRIPTION
**Pull request by soenkekueper**
# PR Content
This PR implements a new modbus binding for openHAB, that integrates Sungrow inverters.
With this binding its very easy to monitor the current values for PV production within the smart home system - without the need to worry about modbus registers, types and transformations. It's implemented along the Sungrow specification "Communication Protocol of Residential Hybrid Inverter V1.0.23",
which can be found here: https://github.com/bohdan-s/SunGather/issues/36.

The binding adds a new thing type "sungrow inverter" that uses the default modbus bridges and needs no further configuration.

# ToDos
~~- Check provided values for battery information~~
~~- complete documentation~~
- Test compatibility with other sungrow inverters.

This binding was already mentioned within thread https://community.openhab.org/t/sungrow-inverter-binding-or-modbus-config/139360/17.

### CODE REVIEW

File: CODEOWNERS [hunk 1]
```diff
@@ -209,6 +209,7 @@
 /bundles/org.openhab.binding.modbus.sbc/ @fwolter
 /bundles/org.openhab.binding.modbus.stiebeleltron/ @pail23
 /bundles/org.openhab.binding.modbus.studer/ @giovannimirulla
+/bundles/org.openhab.binding.modbus.sungrow/ @soenkekueper@gmx.de
```
Comment (jlaur) [inline CODEOWNERS:212]: This is expected to be GitHub username, not e-mail address.

File: CODEOWNERS
Comment (jlaur) [inline CODEOWNERS:1]: It seems like changes from other PR's were reverted?

File: CODEOWNERS (hunk 1 above)
Comment (soenkekueper) [reply CODEOWNERS]: fixed

File: CODEOWNERS
Comment (soenkekueper) [reply CODEOWNERS:1]: Hm, sorry don't know what happend, i've used the maven task for skeleton creation - maybe i've not been up to date and made a mistake on rebasing.
Is fixed.

File: bundles/org.openhab.binding.modbus.sungrow/pom.xml [hunk 2]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>4.0.0-SNAPSHOT</version>
```
Comment (wborn) [inline bundles/org.openhab.binding.modbus.sungrow/pom.xml:10]:
```suggestion
    <version>4.1.0-SNAPSHOT</version>
```

File: bundles/org.openhab.binding.modbus.sungrow/pom.xml [hunk 3]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>4.1.0-SNAPSHOT</version>
```
Comment (wborn) [inline bundles/org.openhab.binding.modbus.sungrow/pom.xml:10]:
```suggestion
    <version>4.2.0-SNAPSHOT</version>
```

**Review by lsiepel — CHANGES_REQUESTED** [review]
Looks very solid. Just some small comments you might want to look at.

Edit: Also some javadoc are incomplete like the public methods in SungrowInverterRegisters.java lack the parameter and return.

File: bundles/org.openhab.binding.modbus.sungrow/README.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
+# Modbus Sungrow Binding
+
+<img src="./doc/sungrow_logo_25_pc.png" alt="Sungrow logo" align="right"/>
```
[the remark this answers is not in the document]
Comment (soenkekueper) [reply bundles/org.openhab.binding.modbus.sungrow/README.md]: Removed.

File: bundles/org.openhab.binding.modbus.sungrow/README.md [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+## Thing Configuration
+
+Once you've configured the Modbus TCP Slave or Modbus Serial Slave as Bridge you can configure the Sungrow inverter thing.
+You just have to select the configured bridge and optional configure the polling interval. 
+
+### `sungrowInverter` Thing Configuration
+
+| Name            | Type    | Description                          | Default | Required | Advanced |
+|-----------------|---------|--------------------------------------|---------|----------|----------|
+| pollInterval    | integer | Interval the device is polled in ms.  | 5000    | no       | no       |
```
[the remark this answers is not in the document]
Comment (soenkekueper) [reply bundles/org.openhab.binding.modbus.sungrow/README.md]: fixed.

File: bundles/org.openhab.binding.modbus.sungrow/src/main/java/org/openhab/binding/modbus/sungrow/internal/SungrowInverterHandler.java [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+                int index = channel.getRegisterNumber() - firstRegister;
+
+                ModbusBitUtilities.extractStateFromRegisters(registers, index, channel.getType())
+                        .map(channel::createState).ifPresent(v -> updateState(createChannelUid(channel), v));
+            }
+        });
+    }
+
+    private void readError(AsyncModbusFailure<ModbusReadRequestBlueprint> error) {
+        this.logger.error("Failed to get modbus data", error.getCause());
```
[the remark this answers is not in the document]
Comment (soenkekueper) [reply bundles/org.openhab.binding.modbus.sungrow/src/main/java/org/openhab/binding/modbus/sungrow/internal/SungrowInverterHandler.java]: I',m not really sure (its a long time ago) so as usually stack traces are usefull for failure-analysis, ive reduced it to debug, so it sould not blow up the log, and is available if needed.

File: bundles/org.openhab.binding.modbus.sungrow/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (soenkekueper) [reply bundles/org.openhab.binding.modbus.sungrow/README.md]: Fixed.

File: bundles/org.openhab.binding.modbus.sungrow/src/main/resources/OH-INF/i18n/sungrow_de.properties [hunk 7]
```diff
@@ -0,0 +1,64 @@
+# thing types
```
[the remark this answers is not in the document]
Comment (soenkekueper) [reply bundles/org.openhab.binding.modbus.sungrow/src/main/resources/OH-INF/i18n/sungrow_de.properties]: Removed.

File: bundles/org.openhab.binding.modbus.sungrow/src/main/resources/OH-INF/thing/thing-types.xml [hunk 8]
```diff
@@ -0,0 +1,527 @@
...
+			<channel id="sg-phase-c-voltage" typeId="sg-phase-c-voltage"/>
+			<channel id="sg-phase-a-current" typeId="sg-phase-a-current"/>
+			<channel id="sg-phase-b-current" typeId="sg-phase-b-current"/>
+			<channel id="sg-phase-c-current" typeId="sg-phase-c-current"/>
+			<channel id="sg-reactive-power" typeId="sg-reactive-power"/>
+			<channel id="sg-grid-frequency" typeId="sg-grid-frequency"/>
+			<channel id="sg-daily-pv-generation" typeId="sg-daily-pv-generation"/>
+			<channel id="sg-total-pv-generation" typeId="sg-total-pv-generation"/>
+			<channel id="sg-power-factor" typeId="sg-power-factor"/>
+			<!-- Not working
```
[the remark this answers is not in the document]
Comment (soenkekueper) [reply bundles/org.openhab.binding.modbus.sungrow/src/main/resources/OH-INF/thing/thing-types.xml]: Removed.

File: bundles/org.openhab.binding.modbus.sungrow/src/main/resources/OH-INF/thing/thing-types.xml [hunk 9]
```diff
@@ -0,0 +1,527 @@
...
+		<label>Grid Information</label>
+		<channels>
+			<channel id="sg-daily-export-energy" typeId="sg-daily-export-energy"/>
+			<channel id="sg-total-export-energy" typeId="sg-total-export-energy"/>
+			<channel id="sg-daily-import-energy" typeId="sg-daily-import-energy"/>
+			<channel id="sg-total-import-energy" typeId="sg-total-import-energy"/>
+			<channel id="sg-daily-export-power-from-pv" typeId="sg-daily-export-power-from-pv"/>
+			<channel id="sg-total-export-energy-from-pv" typeId="sg-total-export-energy-from-pv"/>
+			<channel id="sg-export-power" typeId="sg-export-power"/>
+			<!-- not working
```
[the remark this answers is not in the document]
Comment (soenkekueper) [reply bundles/org.openhab.binding.modbus.sungrow/src/main/resources/OH-INF/thing/thing-types.xml]: Removed too

**Comment by soenkekueper** [comment]
Thanks for your review @lsiepel.

**Review by lsiepel — CHANGES_REQUESTED** [review]
Some more comments, i also noticed that there are some todo's in the start post. Are they done?

File: bundles/org.openhab.binding.modbus.sungrow/README.md [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+Number:Energy daily_discharge_energy "Daily Battery Discharge Energy" (batteryInformation) ["Measurement", "Energy"] {channel="modbus:sungrow-inverter:sungrowBridge:sungrowInverter:sg-battery-information#sg-daily-battery-discharge-energy"}
+
+// Grid information
+Number:Power export_power "Export Power" (gridInformation) ["Measurement", "Power"] {channel="modbus:sungrow-inverter:sungrowBridge:sungrowInverter:sg-grid-information#sg-export-power"}
+Number:Energy daily_export_energy "Daily Export Energy" (gridInformation) ["Measurement", "Energy"] {channel="modbus:sungrow-inverter:sungrowBridge:sungrowInverter:sg-grid-information#sg-daily-export-energy"}
+
+// Load information
+Number:Power load_power "Load Power" (loadInformation) ["Measurement", "Power"] {channel="modbus:sungrow-inverter:sungrowBridge:sungrowInverter:sg-load-information#sg-load-power"}
+Number:Energy daily_direct_energy_consumption "Daily Direct Energy Consumption" (loadInformation) ["Measurement", "Energy"] {channel="modbus:sungrow-inverter:sungrowBridge:sungrowInverter:sg-load-information#sg-daily-direct-energy-consumption"}
+```
```
[the remark this answers is not in the document]
Comment (soenkekueper) [reply bundles/org.openhab.binding.modbus.sungrow/README.md]: Sorry i don't have a sitemap example, a'm only using the main ui.
So i've created a widget: https://community.openhab.org/t/pv-status-widget-for-sungrow-inverters/152120
Is a sitemap example required?

File: bundles/org.openhab.binding.modbus.sungrow/README.md [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+You just have to select the configured bridge and optional configure the polling interval. 
+
+### `sungrowInverter` Thing Configuration
+
+| Name         | Type    | Description                          | Default | Required | Advanced |
+|--------------|---------|--------------------------------------|---------|----------|----------|
+| pollInterval | integer | Interval the device is polled in ms. | 5000    | yes      | no       |
+
+## Channels
+
```
[the remark this answers is not in the document]
Comment (soenkekueper) [reply bundles/org.openhab.binding.modbus.sungrow/README.md:59]: done :)

**Comment by soenkekueper** [comment]
> Some more comments, i also noticed that there are some todo's in the start post. Are they done?

Yes, battery information is working and documentation is updated.
As i've only one inverter, i'm not able to test more. But on forum there are hints on others, that are working.

**Comment by soenkekueper** [comment]
@lsiepel Is everything ok now, or do you have any further comments / questions?

File: bundles/org.openhab.binding.modbus.sungrow/README.md [hunk 12]
```diff
@@ -0,0 +1,104 @@
+# Modbus Sungrow Binding
+
+This binding integrates the sungrow inverters into openHAB.
+It is based on the Sungrow specification "Communication Protocol of Residential Hybrid Inverter V1.0.23",
+which can be found here: https://github.com/bohdan-s/SunGather/issues/36.
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.modbus.sungrow/README.md:5]: Marked as resolved, but i don;t see the change, did you push your changes?

**Comment by lsiepel** [comment]
> @lsiepel Is everything ok now, or do you have any further comments / questions?

Some comments are marked as resolved, but i don;t see any change or comment. Can you check all of them and add a comment like 'fxied' or the reason why it is not canged?

File: bundles/org.openhab.binding.modbus.sungrow/pom.xml (hunk 2 above)
Comment (soenkekueper) [reply bundles/org.openhab.binding.modbus.sungrow/pom.xml]: fixed. (already updated to 4.2.0

File: bundles/org.openhab.binding.modbus.sungrow/README.md [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+## Channels
+
+The `sungrowInverter` thing has channels that serve the current state of the sungrow inverter,
+as you are used to from the iSolareCloud Website and App.
+
+## Configuration Example
+
+This example shows how to configure a sungrow inverter connected via modbus and uses the most common channels.
+
+sungrow.things
```
[the remark this answers is not in the document]
Comment (soenkekueper) [reply bundles/org.openhab.binding.modbus.sungrow/README.md:112]: fixed.

File: bundles/org.openhab.binding.modbus.sungrow/README.md [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+
+```java
+Bridge modbus:tcp:sungrowBridge [ host="10.0.0.2", port=502, id=1, enableDiscovery=false ] {
+    Thing modbus:sungrow-inverter:sungrowBridge:sungrowInverter "Sungrow Inverter" [ pollInterval=5000 ]
+}
+```
+
+sungrow.items
```
[the remark this answers is not in the document]
Comment (soenkekueper) [reply bundles/org.openhab.binding.modbus.sungrow/README.md:120]: fixed.

File: bundles/org.openhab.binding.modbus.sungrow/README.md [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+### `sungrowInverter` Thing Configuration
+
+| Name         | Type    | Description                          | Default | Required | Advanced |
+|--------------|---------|--------------------------------------|---------|----------|----------|
+| pollInterval | integer | Interval the device is polled in ms. | 5000    | yes      | no       |
+
+## Channels
+
+The `sungrowInverter` thing has channels that serve the current state of the sungrow inverter,
+as you are used to from the iSolareCloud Website and App.
```
[the remark this answers is not in the document]
Comment (soenkekueper) [reply bundles/org.openhab.binding.modbus.sungrow/README.md:61]: fixed.

File: bundles/org.openhab.binding.modbus.sungrow/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (soenkekueper) [reply bundles/org.openhab.binding.modbus.sungrow/README.md:5]: fixed.

File: bundles/org.openhab.binding.modbus.sungrow/README.md [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+- SH5.0RT
+- SH6.0RT
+- SH8.0RT
+- SH10RT
+
+## Supported Things
+
+The binding supports only one thing:
+
+- `sungrowInverter`: The sungrow inverter
```
[the remark this answers is not in the document]
Comment (soenkekueper) [reply bundles/org.openhab.binding.modbus.sungrow/README.md:32]: fixed.

**Comment by soenkekueper** [comment]
Sorry i don't know what happend - maybe some rebase/merging error on my side. sorry for that.
Now all remarks should be resolved - exept the sitemap - i don't have one as i'm only using the main-ui.
Is it required?

File: bundles/org.openhab.binding.modbus.sungrow/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.modbus.sungrow/README.md:148]: I have added the example to your PR.

**Review by lsiepel — APPROVED** [review]
Thanks for providing this new binding. LGTM

**Pull request merged**